### ISSUE TITLE
Folds on bound threads

### ISSUE THREAD

**Issue by shlok** [composewell/streamly#552]
Does the current API allow me to somehow create a `Fold` which is guaranteed to occur on a bound thread (when eventually executed with `fold`)?

Alternatively, what about a `fold` alternative such as `foldBound`? (This latter alternative is not ideal, as it places the burden of thinking about bound vs. unbound on the user of the `Fold`.)

**Comment by harendra-kumar** [comment]
No, currently there is no way to control whether some piece of code runs on a bound thread, in fact all threads are unbound. We do not use forkOS to create a thread.

What is the use case for this?

**Comment by shlok** [comment]
I aim to create a `Fold m (ByteString, ByteString) ()` that represents an LMDB write transaction. Write transactions in LMDB are not allowed to be used on multiple OS threads. Hence my inquiry.

With the existing API, I see two ways: 1) Ask users of the `Fold` to wrap their usage within something like `asyncBound` or `forkOS`. (2) Accumulate the bytestrings into an intermediate data structure (e.g., `Seq` or `Vector`) and then write them out when the fold ends.

(2) has better user experience (no need to think about bound threads), but worse performance.

**Comment by harendra-kumar** [comment]
To resolve this, we can have proper support for bound threads in streamly. Since boundThreads is a specific use case I think its best left to the programmer to decide whether to use a bound thread. Also, bound threads could be very expensive if we use them liberally like the regular threads.

It should not be difficult. Take a look at [maxThreads here](https://github.com/composewell/streamly/blob/f80a75fcc759942e68e0ce1defb14871938e2455/src/Streamly/Internal/Data/Stream/Combinators.hs#L55). On the same lines we can have `boundThreads` and `unboundThreads` combinators. `boundThreads` would set a flag in the state and unboundThreads would clear it. Based on this flag, in `pushWorker` or `pushWorkerPar`, we can decide to forkOS instead of fork.

Once that is done, you can use `boundThreads`  with `foldAsync` to run your fold on a bound thread. 
Would you like to have a go at this?

**Comment by shlok** [comment]
Excellent. Sure, I’m happy to give it a try, possibly if not probably in the coming week or two. At first glance, it seems that `doFork` would end up with an additional `Boolean` “bound” parameter (populated from the new state flag that you proposed); does that sound right? I also wonder which benchmarks are relevant for running afterwards.

**Comment by harendra-kumar** [comment]
I guess you could store the state in the `SVar` when creating it. Wherever `doFork` is invoked we have access to the `SVar` and can get the flag from there and pass it to `doFork`. I am just quickly guessing, when implementing you can decide whatever makes better sense.

See this https://github.com/composewell/streamly/blob/master/benchmark/README.md for running benchmarks. You can run the `prelude_concurrent_grp` group of benchmarks, initially you can just run `Prelude.Async` single benchmark suite. You can add a benchmark(s) with bound threads setting set to true in `benchmark/Streamly/Benchmark/Prelude/Async.hs `.

**Comment by shlok** [comment]
I’ll close this now, as this is being addressed by my new PR: https://github.com/composewell/streamly/pull/2400
